controllers/operatorpolicy_controller.go

```diff
@@ -133,9 +146,96 @@ func (r *OperatorPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 
 	// FUTURE: more resource checks
 
+	// handle catalogsource
+	if err := r.handleCatalogSource(ctx, policy); err != nil {
+		OpLog.Error(err, "Error handling CatalogSource")
+
+		return reconcile.Result{}, err
+	}
+
 	return reconcile.Result{}, nil
 }
 
+func (r *OperatorPolicyReconciler) handleCatalogSource(
+	ctx context.Context,
+	policy *policyv1beta1.OperatorPolicy,
+) error {
+	watcher := opPolIdentifier(policy.Namespace, policy.Name)
+
+	desiredSub, err := buildSubscription(policy)
+	if err != nil {
+		return fmt.Errorf("error building Subscription: %w", err)
+	}
+
+	// Retrieve Subscription
+	foundSub, err := r.DynamicWatcher.Get(watcher, subscriptionGVK, desiredSub.Namespace, desiredSub.Name)
+	if err != nil {
+		return fmt.Errorf("error getting Subscription: %w", err)
+	}
+
+	if foundSub == nil {
+		// Subscription not found, check the CatalogSource object
+		foundCatalogSrc, err := r.DynamicWatcher.Get(watcher, catalogSrcGVK,
+			desiredSub.Spec.CatalogSourceNamespace, desiredSub.Spec.CatalogSource)
+		if err != nil {
+			return fmt.Errorf("error getting CatalogSource: %w", err)
+		}
+
+		// Conditionally emit health event using the defined status template
+		isMissing := foundCatalogSrc == nil
+
+		err = r.updateStatus(ctx, policy, catalogSourceFindCond(isMissing), catalogSourceObj(desiredSub, isMissing))
+		if err != nil {
+			return fmt.Errorf("error updating the status for CatalogSource when Subscription DNE on cluster: %w", err)
+		}
+
+		return nil
+	}
+
+	condList, found, err := unstructured.NestedSlice(foundSub.Object, "status", "conditions")
+	if !found {
+		return fmt.Errorf("field does not exist in object: %w", err)
+	}
+
+	if err != nil {
+		return fmt.Errorf("error decoding object into array type: %w", err)
+	}
+
+	// Iterate over all the conditions to determine the state of the CatalogSource
+	for _, condition := range condList {
+		condMap, ok := condition.(map[string]interface{})
+		if !ok {
+			return fmt.Errorf("conditions field is of the wrong type")
+		}
+
+		if fmt.Sprintf("%v", condMap["type"]) == catalogSrcConditionType {
+			if fmt.Sprintf("%v", condMap["status"]) == string(metav1.ConditionFalse) {
+				// Emit healthy event using the defined status template
+				err := r.updateStatus(ctx, policy, catalogSourceFindCond(false), catalogSourceObj(desiredSub, false))
```

Comment:
So if the Subscription says that the CatalogSource isn't unhealthy, this just reports it as found... Does the Subscription have a reason and message in this case, can we report that instead?

I'm thinking it would be more complete if, when the condition is found, it copies over all of the information from that condition, like it currently does in the `else` case below.

Reply:
I think originally I didn't have separate logic to handle when the `Subscription` did not exist so we couldn't read from the conditions field. With the most recent changes I think I can change this logic to inherit the `Subscription`'s condition instead.